### DISCUSSION TITLE
Struggling to use Image Automation

### DISCUSSION THREAD

**Discussion by hamishforbes** [fluxcd/flux2#2562] [Q&A]
I'm having real difficulties getting the Fluxv2 Image Automation process to fit into my configuration workflows.

Currently I am using Fluxv1 with the legacy shell/command based manifest generation feature.
I use [Tanka](https://tanka.dev/) to generate Kubernetes yaml which gets piped into Fluxv1 and applied to the cluster.
Flux commits a patch file back to the repo, Tanka doesn't need to know anything about this and it all works perfectly.
Except for performance, so I'm trying to move to Fluxv2

I now have a workflow where I commit Tanka Jsonnet code to `master` in my repo, github actions uses [tk export](https://tanka.dev/exporting) to generate plain Kubernetes yaml for every environment/cluster combination into an alternate branch (`deploy/master`) alongside an extra 'environment' containing all the Flux `Kustomization` resources relevant to the cluster

So my deploy branch looks something like...

```
environments-staging-euw1-base/
                              ../deployment-abc.yaml
                              ../service-abc.yaml
                              ../configmap-abc.yaml
                              ../deployment-xyz.yaml
                              ../service-xyz.yaml
                              ../configmap-xyz.yaml
environments-staging-euw1-foo/
                              ../bar.yaml
environments-staging-euw1-flux/
                              ../kustomization-base.yaml
                              ../kustomization-foo.yaml

flux/staging-euw1/flux-system/
                              ../gotk-components.yaml
                              ../gotk-sync.yaml
                              ../kustomization.yaml
```

So `base` and `foo` are Tanka environments for cluster `staging-euw1`.
The `flux/staging-euw1/flux-system/` is from the Fluxv2 bootstrapping (via Terraform).
My terraform bootstrapping also adds has 1 extra ` Kustomization` to load `environments-staging-euw1-flux`.

Tanka can export to a directory with other files in it.
It won't export over the top of an existing file/resource.
And it can't generate comments.

This makes merging Flux's automated image updates back in difficult.

I can't comment the deployment yamls directly,
even if I did that in an extra step after the Tanka export, the next time I commit to `master` and re-export the image updates would be overridden.
Possibly I could setup something to extract the current images, stash them and then re-apply after exporting. Pretty convoluted though.

I can't add it to the `images` array of the `Kustomization` resources in `environments-staging-euw1-flux/` for the same reason, they'll be wiped out next time Tanka exports.

I think I could add the images to an `images` array in a `kustomization.yaml` in each environment (e.g. `environments-staging-euw1-foo/kustomization.yaml`)
But I would have to a) know (or determine) all the images I want to auto update and list them there, with comments which means I can't use Tanka to generate these
b) list every resource yaml generated by Tanka in that environment as well, this isn't so difficult I can just `ls` the directory after export, but still a bit annoying

Does anyone have any other options they can think of?
At this point the extra level of complexity I'm having to add to support FluxV2 is making me seriously consider ditching Flux entirely and using Tanka to directly apply config to the cluster.
Figuring something else out for automatic image deployments.

**Comment by kingdonb** [comment]
If you are trying to use image update automation on a generated file as you have seen there will be issues.

Flux has to locate a kyaml setter comment (the marker) which indicates where to write an update for Image Automation Controller. If the file is generated ahead of time by some CI process, (1) this comment is not likely to be preserved, and (2) on the next generate Flux's updates are going to be overwritten rather than merged.

I would suggest keeping a ConfigMap where the version to deploy lives, somewhere outside of the generated manifest directory, and using that configmap in a Variable Substitution so that the generated manifests can contain `${VERSION}` and Flux can substitute the value into the manifests at deploy time, based on the value in the configmap.

Since you are doing manifest generation, I also want to check if you have seen the [Manifest Generation use case](https://fluxcd.io/docs/use-cases/gh-actions-manifest-generation/) guide which covers this topic in detail - there is an example of doing this in particular there: https://fluxcd.io/docs/use-cases/gh-actions-manifest-generation/#jsonnet-configmap-with-envsubst
